pkg/xds/bootstrap/template_v3.go

```diff
@@ -326,3 +313,58 @@ func clusterTypeFromHost(host string) envoy_cluster_v3.Cluster_DiscoveryType {
 	}
 	return envoy_cluster_v3.Cluster_STRICT_DNS
 }
+
+func getGrpcServices(params configParameters, useTokenPath bool) []*envoy_core_v3.GrpcService {
+	var grpcSerivces []*envoy_core_v3.GrpcService
+	if useTokenPath && params.DataplaneTokenPath != "" {
+		googleGrpcService := &envoy_core_v3.GrpcService{
+			TargetSpecifier: &envoy_core_v3.GrpcService_GoogleGrpc_{
+				GoogleGrpc: &envoy_core_v3.GrpcService_GoogleGrpc{
+					TargetUri:              fmt.Sprintf("%s:%d", params.XdsHost, params.XdsPort),
+					StatPrefix:             "ads",
+					CredentialsFactoryName: "envoy.grpc_credentials.file_based_metadata",
+					CallCredentials: []*envoy_core_v3.GrpcService_GoogleGrpc_CallCredentials{
+						{
+							CredentialSpecifier: &envoy_core_v3.GrpcService_GoogleGrpc_CallCredentials_FromPlugin{
+								FromPlugin: &envoy_core_v3.GrpcService_GoogleGrpc_CallCredentials_MetadataCredentialsFromPlugin{
+									Name: "envoy.grpc_credentials.file_based_metadata",
+									ConfigType: &envoy_core_v3.GrpcService_GoogleGrpc_CallCredentials_MetadataCredentialsFromPlugin_TypedConfig{
+										TypedConfig: util_proto.MustMarshalAny(&envoy_grpc_credentials_v3.FileBasedMetadataConfig{
+											SecretData: &envoy_core_v3.DataSource{
+												Specifier: &envoy_core_v3.DataSource_Filename{Filename: params.DataplaneTokenPath},
+											},
+										}),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		if params.CertBytes != nil {
+			googleGrpcService.GetGoogleGrpc().ChannelCredentials = &envoy_core_v3.GrpcService_GoogleGrpc_ChannelCredentials{
+				CredentialSpecifier: &envoy_core_v3.GrpcService_GoogleGrpc_ChannelCredentials_SslCredentials{
+					SslCredentials: &envoy_core_v3.GrpcService_GoogleGrpc_SslCredentials{
+						RootCerts: &envoy_core_v3.DataSource{
+							Specifier: &envoy_core_v3.DataSource_InlineBytes{
+								InlineBytes: params.CertBytes,
+							},
+						},
+					},
+				},
+			}
+		}
+		grpcSerivces = append(grpcSerivces, googleGrpcService)
```

Comment:
correct me if I'm wrong, but I think in this case, we don't need `ads_cluster`, right? I can't see a condition when we create `ads_cluster`